report/formatted.go

```diff
@@ -0,0 +1,191 @@
+		src:  src,
+		logf: logf,
+	}
+}
+
+// Source returns the target source.
+func (rep *LogfReporter) Source() source.DataSource {
+	return rep.src
+}
+
+// Start clears any stop flag, and starts watching the data source.
+func (rep *LogfReporter) Start() error {
+	var err error
+	rep.stopped = false
+	if isrc, ok := rep.src.(source.ItemDataSource); ok {
+		err = isrc.WatchItems("text", rep)
+	} else {
+		err = errRawSource
+	}
+	if err != nil {
+		rep.stopped = true
+	}
+	return err
+}
+
+// Stop sets a flag internally so that the next HandleItem(s) will return an
+// error, removing the watcher resource.
+func (rep *LogfReporter) Stop() {
+	rep.stopped = true
+}
+
+// HandleItem outputs the item to the logging function with a source-name
+// prefix.
+func (rep *LogfReporter) HandleItem(item []byte) error {
+	if rep.stopped {
+		return errReporterClosed
+	}
+	rep.logf("%s: %s", rep.src.Name(), item)
+	return nil
+}
+
+// HandleItems outputs all items to the logging function with a source-name
+// prefix on each item.
+func (rep *LogfReporter) HandleItems(items [][]byte) error {
+	if rep.stopped {
+		return errReporterClosed
+	}
+	name := rep.src.Name()
+	for _, item := range items {
+		rep.logf("%s: %s", name, item)
+	}
+	return nil
+}
+
+// PrintfReporter is a FormattedReporter that targets a log formatting function.
+type PrintfReporter struct {
+	src     source.DataSource
+	printf  func(format string, args ...interface{}) (int, error)
+	stopped bool
+}
+
+// NewPrintfReporter creates a new FormattedReporter around a raw
+// fmt.Printf-family formatting function.  The formatting function is expected
+// to return a number and error in package fmt style.  NewPrintfReporter will
+// append a newline to passed format strings, since the print formatting
+// function is expected to not do so.
+func NewPrintfReporter(
+	src source.DataSource,
+	printf func(format string, args ...interface{}) (int, error),
+) *PrintfReporter {
+	return &PrintfReporter{
+		src:    src,
+		printf: printf,
+	}
+}
+
+// Source returns the target source.
+func (rep *PrintfReporter) Source() source.DataSource {
+	return rep.src
+}
+
+// Start clears any stop flag, and starts watching the data source.
+func (rep *PrintfReporter) Start() error {
+	var err error
+	rep.stopped = false
+	if isrc, ok := rep.src.(source.ItemDataSource); ok {
+		err = isrc.WatchItems("text", rep)
+	} else {
+		err = errRawSource
+	}
+	if err != nil {
+		rep.stopped = true
+	}
+	return err
+}
+
+// Stop sets a flag internally so that the next HandleItem(s) will return an
+// error, removing the watcher resource.
+func (rep *PrintfReporter) Stop() {
+	rep.stopped = true
+}
+
+// HandleItem outputs the item to the printf function with a source-name
+// prefix and trailing newline.
+func (rep *PrintfReporter) HandleItem(item []byte) error {
+	if rep.stopped {
+		return errReporterClosed
+	}
+	if _, err := rep.printf("%s: %s\n", rep.src.Name(), item); err != nil {
+		rep.stopped = true
+		return err
```

Comment:
Instead of copying this in HandleItems, consider adding a helper `handleItems` which doesn't check `rep.stopped` and call it from both, `HandleItem` and `HandleItems`.